File: javascript/romanToInt.js
/**
 * For example, 2 is written as II in Roman numeral, just two one's added together. 
 * 12 is written as XII, which is simply X + II. 
 * The number 27 is written as XXVII, which is XX + V + II.
 * @param {string} s
 * @return {number}
 */
 var romanToInt = function(s) {
    //  function output
    let output = 0;

    // create a map to index roman keys with values
    const map = new Map();
    map.set('I', 1)
    map.set('V', 5)
    map.set('X', 10)
    map.set('L', 50)
    map.set('C', 100)
    map.set('D', 500)
    map.set('M', 1000)
    map.set('IV', 4)
    map.set('IX', 9)
    map.set('XL', 40)
    map.set('XC', 90)
    map.set('CD', 400)
    map.set('CM', 900)

    // iterate through the string of roman numerals
    for (let current = 0; current < s.length; current++) {
        // if there exists a map key with multiple roman numerals, add to output and increment by 2
        if (map.has(s[current] + s[current+1])) {
            output += map.get(s[current] + s[current+1])
            current++;
        }
        // if not, just get roman numeral value and add to output
        else {
            output += map.get(s[current])
        }        
    }
    return output
};

console.log(romanToInt('MCMXCIV'))

File: python/roman_to_int.py
'''
https://leetcode.com/problems/roman-to-integer/

Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.
'''
class Solution:
    def romanToInt(self, s: str) -> int:
        output = 0
        table = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000,
            'IV': 4,
            'IX': 9,
            'XL': 40,
            'XC': 90,
            'CD': 400,
            'CM': 900
        }
        # since we need to increment x by 1 within the loop, we must create an itorator
        # if not, then x will reset after iterating within the loop.
        iterator = iter(range(0, len(s)))
        # iterate through list
        for x in iterator:
            # if there exists a dicitonary key with multiple roman numerals, add to output and skip next itoration
            if x+1 < len(s) and s[x] + s[x+1] in table:
                output += table[s[x] + s[x+1]]
                next(iterator, None)
            else:
                output += table[s[x]]
        return output


test = Solution()
print(test.romanToInt('MCMXCIV'))